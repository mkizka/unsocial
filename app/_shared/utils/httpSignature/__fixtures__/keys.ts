const publickKey = `-----BEGIN PUBLIC KEY-----
MIIBHjANBgkqhkiG9w0BAQEFAAOCAQsAMIIBBgKB/gDJ5rcs3GJ+7YNOZ2qYgBMX
77suMZmpXNo1YQHyi4ZQkbvojsv7fuDUwjSR7CL78xN88pnTgJTrJfMjeH3LCV/R
BvcsRQNK5QRLvD9ieGLo+1cNycyr1WScTz3RmfkCAGhe4vhmBIlne9w1ItwrFAnP
qmFuAswgEXH24WMLGocPo/fskUoiqtPzqw73r2FvXHv9Qv588lswy3NsL7iANExz
fYzWy8nsa05UH1SrEJ9uT+Bh6ffupJtPuZ/lDSMl2qmHs4kQX67V1amKHl6U6eyx
/CLySZ7hI/SZfXE4ppyL/l+se8ti1I0OcUVMSXjyIN1Z2SNHPHu2QGx2/EGRAgMB
AAE=
-----END PUBLIC KEY-----
`;

const privateKey = `-----BEGIN PRIVATE KEY-----
MIIEqAIBADANBgkqhkiG9w0BAQEFAASCBJIwggSOAgEAAoH+AMnmtyzcYn7tg05n
apiAExfvuy4xmalc2jVhAfKLhlCRu+iOy/t+4NTCNJHsIvvzE3zymdOAlOsl8yN4
fcsJX9EG9yxFA0rlBEu8P2J4Yuj7Vw3JzKvVZJxPPdGZ+QIAaF7i+GYEiWd73DUi
3CsUCc+qYW4CzCARcfbhYwsahw+j9+yRSiKq0/OrDvevYW9ce/1C/nzyWzDLc2wv
uIA0THN9jNbLyexrTlQfVKsQn25P4GHp9+6km0+5n+UNIyXaqYeziRBfrtXVqYoe
XpTp7LH8IvJJnuEj9Jl9cTimnIv+X6x7y2LUjQ5xRUxJePIg3VnZI0c8e7ZAbHb8
QZECAwEAAQKB/WbVoE61IWXMy6hPfDeYZCE49XmdwcpFhM7TTg6fihXXZIIa2UKL
aTB0ihoYz+jL8g0d88OKJk9u0NwsLsYPTEJG7fYaYIbxBrMcepxI/Pf1rG1Vb23G
Dqyen7n2VqHyWzaZHDO589JFf++AhetbFXQ0L9O8kQN0XzuDqNKq/Dk69LwBlnxX
gK5BROV03coTwwBZGkX+WaSBWmRxurWaN4Mskpsv0ITXi8CD9dyIvSoFlig+UV1Q
2YqLr9O7QUkEnEH2OwaBowilOfEgIokY5f0TseuGdFQ1Pue+l39wiq5mWUjDWpzY
ja2wc2JeL70ph8HRZFPPszclYtTA1RECfw7DL5tIJUUdjIUkuEP0jlMeojTgW5bm
aZzw1nv694ftgLqFclkIfvU6KVIOdVfSWCcCedFJnIm31KmBrQJyCUV9cYg/Mdge
/EiWWATjl+FkIzkHOVw35bpBjQKmTV45Nmf9zzuIkdRKtRqettBimLVRfShqmbXh
h2xpO19aqYcCfw2tOt+j3hv+vt/sQR19oTT0leydNZpsRbvFwtdjLLZ3mCym04l+
XR8SVxps+jHJDGNZyvT3ps0eLL9FtF6rwDxpU46LN9y1EwtT+ZT7TtYpTX3KbrMt
yzsegf6cACErFvDLV4rGeMfZfhUHhxG3btwywymJ+hT9hPhjbp0zoicCfwL/+z6t
DJXdCV9NNPzqR7nwcboZFcO2kyPHdW7Ac2j2BMDiuXgKV/JxnVAFjCcKoctbn3G+
acc6ykagCY6DqgqrEVVEslZfBZmio8Rqk8TD2ryOEUQCP/b3tzYgoXxSWBIN6/0H
JctkWyVLt47lWm8OXkbGUvI5IJIRWHVqMSUCfwEVe5cB27aISARjKv31WPNjdwZb
p0Xn0tcA42JCn+aOL+dKCXXaJ/MwdurHPET51BuDoEwt3HBhNdaq5ZKjNgcTqeWg
Lvs09p9tNliSWjsq5tNTWdaliHfVASR2pKP8UEMhOnzrIeNk31M3V4MV8JlPBJyk
LMfBF+q/pqKa1v0Cfw3+go4Vmz0JyVex74CpdqNXrjvHU873ypjSOemeoRP5T8qz
Pw9+4NLFLb7E9vZTCLGyBW4OCKb0fyrbkWDAh/9/gv8hgvtM71Fgj8iFR0FrgG4b
s+7Y1gr9CevUwUnFFtagFM2oEnOYydaqxq5BvXgFdoH6Whr9Gl03RQOl7QU=
-----END PRIVATE KEY-----
`;

export const mockedKeys = { publickKey, privateKey };
