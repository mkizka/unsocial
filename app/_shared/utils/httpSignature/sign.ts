import crypto from "crypto";

import { env } from "@/_shared/utils/env";

import { createDigest, textOf } from "./utils";

const getSignature = (textToSign: string, privateKey: string) => {
  const sig = crypto.createSign("sha256").update(textToSign).end();
  return sig.sign(privateKey, "base64");
};

export type SignActivityParams = {
  signer: {
    id: string;
    privateKey: string;
  };
  body: string;
  inboxUrl: URL;
  method: string;
};

// TODO: Requestオブジェクトを受け取るようにする
export const signHeaders = ({
  signer,
  body,
  inboxUrl,
  method,
}: SignActivityParams) => {
  const order = [
    "(request-target)",
    "host",
    "date",
    method === "POST" ? "digest" : "accept",
  ];
  const headers: Record<string, string> = {
    host: inboxUrl.host,
    date: new Date().toUTCString(),
    digest: `SHA-256=${createDigest(body)}`,
  };
  if (method === "GET") {
    headers.accept = "application/activity+json";
  }
  const textToSign = textOf({
    method,
    pathname: inboxUrl.pathname,
    headers,
    order,
  });
  const signature = getSignature(textToSign, signer.privateKey);
  return {
    ...headers,
    signature:
      `keyId="https://${env.UNSOCIAL_DOMAIN}/users/${signer.id}/activity#main-key",` +
      `algorithm="rsa-sha256",` +
      `headers="${order.join(" ")}",` +
      `signature="${signature}"`,
  };
};
