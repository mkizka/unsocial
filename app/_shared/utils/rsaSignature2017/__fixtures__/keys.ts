export const rsaKeyPair = {
  private:
    "-----BEGIN RSA PRIVATE KEY-----\nMIIEpAIBAAKCAQEAxiD/F5M/LPKsNeTV1R55dtsQRYqP0835BuxuYyAeTENDwpji\neFZHztrmpPFqnfPgGlW3aDfN84PfC5F+LznB1EmWlfQwWTTTb7VurlMmHOgQx9c+\n2+Yg22cgPGh8qTxUALZFpQ+DHVBU59FXbzpZKeF2go9fqvJvYaxcf5Sq1pu/Y62k\nplCVxmow508agslRvqAzowsHvB/Dp5wETaXiyvXeqCFSEmTDCu3CCnu4HuwwP+H3\nKNSuAPZtcjOlCxrvO50RvzT5RnnRITuTi4oimIJMQoIX6N/RZzgFRofsTBKqbjFs\nZ6hja496yLp/3KJ8egQ2srl4kt3TzHlDfKXPnQIDAQABAoIBAF1mirrjaGHZGMPj\nV8a7MgmM5dPSqVDxBZd09+j06DONdjETf5e5HOrwGNDjGh7SK/BePjL8M7O7k/dR\nJEpBERXxhvK0BeLX22ELB91YR1xTLi33U6O4jf7AU9sGatYhIXygcJ6KfIXiK9SI\nxI4j1IrAxexwiGn/FfDWn1hpjkvSlUQr14YCCuckBXAkAAjqlcrH0xEkcFls4HAo\nwn+vCaxN+tQaP9fmwbBMyILSLgmirFeZPJbtL731h677jED0Gre2PQJc2UzNnWA2\nThKlIqikTbyNAULC9u3+dtSqARD2NWTh8rhnkyiPQ3giYezWnbDX7/gEZx6ZhjmA\nA5Bc0kUCgYEA9P4khD7fBQURD84boR6WsCd6VEpBuDqEzx6HDIUTwkj2VG6ftzz6\nTwCZmhgnRCedutCzKqfH4TCMlXUYoXgg/8lrmwVxWCYt41NWr4cMS3GThXwEBrDN\nHUu+BCEEWmUdaILXlpD7XFz2bOi45cj8jp/UQSZKC9wKIJwT2OLuydMCgYEAzwfV\nzxfxGRf/MfN6nhEGVgJ03YpspIgIkLZQctZOBTPBUXsECYj85cAdPtPwXbHHqx4u\nW63FqHhOBZdf8b8a8zOCTjnxiZAzlTGezeK9jMtQM5+IgEWNf46Zzt+fLnooq4Rc\nGcgsNwp3xuBMKPPiCEXggUlICAEoEAB1e5E8Ks8CgYB6oBZhByMT3SRcF9rme2wL\nQcdOF1jrwIanMO2By75HGcQKINDao6/aCT/YzEwjJf0h+8IiEJbQVT+Ys2Kft4Q0\n43lzv3arG3kOdrFF7D+wJ2oifx94xdf02OuEH46RPFlK7WrBV1IpOfKyMo4Rndck\nAEX6UhLv+rQq/FRG09eXtwKBgQC0MrqsFce5ssRh4u4gNOtR47N6vSiREQJ7iuR9\nAAFKz+B9HljGP+D3zP7nl6uIsdv8Gjn5fEglugkXHsnY7iDz/Cp4EkElHkl3m4gA\n5lFcAf8d38b89w0RMLq5cLTpzFQn7f4IRs4NXmQgy0we/SAJWWDuxy0WIoOPYgbf\n9errywKBgQDC7NPm+2k8Mw05JRKQ+yQRhsEOh3/gDFz28w/DPqmm5Z5YOhVIK5bd\n1mSEklsX8dSt5YObMcVavt6+gWi3lPkznO063BzpnPDCBTbApyaB7B8hlKy/KtbZ\nzwYQC2O+AflQIvjZu3NaAAZ288V9YVz6pys9ycBweOoQuJkovLIsCg==\n-----END RSA PRIVATE KEY-----\n",
  public:
    "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxiD/F5M/LPKsNeTV1R55\ndtsQRYqP0835BuxuYyAeTENDwpjieFZHztrmpPFqnfPgGlW3aDfN84PfC5F+LznB\n1EmWlfQwWTTTb7VurlMmHOgQx9c+2+Yg22cgPGh8qTxUALZFpQ+DHVBU59FXbzpZ\nKeF2go9fqvJvYaxcf5Sq1pu/Y62kplCVxmow508agslRvqAzowsHvB/Dp5wETaXi\nyvXeqCFSEmTDCu3CCnu4HuwwP+H3KNSuAPZtcjOlCxrvO50RvzT5RnnRITuTi4oi\nmIJMQoIX6N/RZzgFRofsTBKqbjFsZ6hja496yLp/3KJ8egQ2srl4kt3TzHlDfKXP\nnQIDAQAB\n-----END PUBLIC KEY-----\n",
};
