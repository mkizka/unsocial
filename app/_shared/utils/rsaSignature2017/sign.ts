import crypto from "crypto";

import { createVerifyData } from "./common";

type SignOptinos = {
  activity: object;
  creator: string;
  domain?: string;
  created?: Date;
  privateKey: string;
};

// 参考:
//   https://github.com/transmute-industries/RsaSignature2017/blob/5cd9f90989e313d5039a4e63cdb7677ae8467b02/src/index.js
//   https://github.com/misskey-dev/misskey/blob/776eea736a2607c62145c01b48a8271dd355f006/packages/backend/src/core/activitypub/LdSignatureService.ts
const sign = async ({
  activity,
  creator,
  domain,
  created,
  privateKey,
}: SignOptinos) => {
  const options = {
    type: "RsaSignature2017",
    creator,
    domain,
    nonce: crypto.randomBytes(16).toString("hex"),
    created: created || new Date().toISOString(),
  };
  if (!domain) {
    delete options.domain;
  }
  const toBeSigned = await createVerifyData(activity, options);
  const signer = crypto.createSign("sha256");
  signer.update(toBeSigned);
  signer.end();
  const signature = signer.sign(privateKey);
  return {
    ...activity,
    signature: {
      ...options,
      signatureValue: signature.toString("base64"),
    },
  };
};
