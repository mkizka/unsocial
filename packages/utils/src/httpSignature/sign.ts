import crypto from "crypto";

import { env } from "../env";
import { textOf } from "./utils";

const createDigest = (activity: object) => {
  return crypto
    .createHash("sha256")
    .update(JSON.stringify(activity)!)
    .digest("base64");
};

const getSignature = (textToSign: string, privateKey: string) => {
  const sig = crypto.createSign("sha256").update(textToSign).end();
  return sig.sign(privateKey, "base64");
};

type Sender = {
  id: string;
  privateKey: string;
};

// TODO: 理解する
// https://docs.joinmastodon.org/spec/security/
export const signActivity = (params: {
  sender: Sender;
  activity: object;
  inboxUrl: URL;
}) => {
  const order = ["(request-target)", "host", "date", "digest"];
  const header = {
    host: params.inboxUrl.host,
    date: new Date().toUTCString(),
    digest: `SHA-256=${createDigest(params.activity)}`,
  };
  const headerToSign = {
    "(request-target)": `post ${params.inboxUrl.pathname}`,
    ...header,
  };
  const textToSign = textOf(headerToSign, order);
  const signature = getSignature(textToSign, params.sender.privateKey);
  return {
    ...header,
    signature:
      `keyId="https://${env.HOST}/users/${params.sender.id}#main-key",` +
      `algorithm="rsa-sha256",` +
      `headers="${order.join(" ")}",` +
      `signature="${signature}"`,
  };
};
