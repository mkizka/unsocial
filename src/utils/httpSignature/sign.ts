import type { AP } from "activitypub-core-types";
import crypto from "crypto";
import { textOf } from "./utils";

const createDigest = (activity: object) => {
  return crypto
    .createHash("sha256")
    .update(JSON.stringify(activity)!)
    .digest("base64");
};

const getSignature = (textToSign: string, privateKey: string) => {
  const sig = crypto.createSign("sha256").update(textToSign).end();
  return sig.sign(privateKey, "base64");
};

// TODO: 理解する
// https://docs.joinmastodon.org/spec/security/
export const signActivity = (
  activity: AP.Activity,
  inboxUrl: URL,
  publicKeyId: string,
  privateKey: string
) => {
  const order = ["(request-target)", "host", "date", "digest"];
  const header = {
    host: inboxUrl.host,
    date: new Date().toUTCString(),
    digest: `SHA-256=${createDigest(activity)}`,
  };
  const headerToSign = {
    "(request-target)": `post ${inboxUrl.pathname}`,
    ...header,
  };
  const textToSign = textOf(headerToSign, order);
  const signature = getSignature(textToSign, privateKey);
  return {
    ...header,
    signature:
      `keyId="${publicKeyId}",` +
      `algorithm="rsa-sha256",` +
      `headers="${order.join(" ")}",` +
      `signature="${signature}"`,
  };
};
