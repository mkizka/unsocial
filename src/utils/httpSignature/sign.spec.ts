import { mockedKeys } from "./fixtures/keys";
import { signActivity } from "./sign";

afterAll(() => {
  jest.useRealTimers();
});

describe("signActivity", () => {
  test.each`
    url                                                      | date            | activity              | description
    ${new URL("https://remote.example.com/inbox")}           | ${"2023-01-01"} | ${{}}                 | ${"default"}
    ${new URL("https://remote1.example.com/inbox")}          | ${"2023-01-01"} | ${{}}                 | ${"url"}
    ${new URL("https://remote.example.com/users/foo/inbox")} | ${"2023-01-01"} | ${{}}                 | ${"path"}
    ${new URL("https://remote.example.com/inbox")}           | ${"2023-01-02"} | ${{}}                 | ${"date"}
    ${new URL("https://remote.example.com/inbox")}           | ${"2023-01-01"} | ${{ type: "Create" }} | ${"activity"}
  `("署名されたヘッダーを返す: $description", ({ url, date, activity }) => {
    // arrange
    jest.useFakeTimers().setSystemTime(new Date(date));
    // act
    const headers = signActivity(
      activity,
      url,
      "https://myhost.example.com/example#main-key",
      mockedKeys.privateKey
    );
    // assert
    expect(headers).toMatchSnapshot();
  });
});
