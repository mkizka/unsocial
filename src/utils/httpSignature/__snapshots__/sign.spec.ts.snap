// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signActivity 署名されたヘッダーを返す: activity 1`] = `
{
  "date": "Sun, 01 Jan 2023 00:00:00 GMT",
  "digest": "SHA-256=JeE18werLvQnEoHViKDam+ZK1D8E27TBC2kIISI7pIY=",
  "host": "remote.example.com",
  "signature": "keyId="https://myhost.example.com/example#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="i335/DHLc9AVBH8vUEyucyC6hfQGdp0iBaD01FWdyrmJYjwr3cyvmzDJPg931z6wK01T2ef6VDq/VrvvR6qfXw8/e3Sx/iGgt34eNs6OS4Oen/qjgXk4Baq4zMdzMiijCXNcS956NG1UMYMZx6h8OgYsqtleoTlaRc/RLdDLiZgLk1uI/uxdyZgrh1dgQkAASMBTxVaWA5lO5me78JOKpRSgb5V/yQQHvHbqnwCMefF0x96wZ8LZjEKq3+8c9Rb2s4F7BQMrt5NnhvEl/Jz7wuF8VRn6zudyv3DQBIGWCuJlDdG/9etcRmtjAFEiI9itcyONIr90++WBy6vrUw=="",
}
`;

exports[`signActivity 署名されたヘッダーを返す: date 1`] = `
{
  "date": "Mon, 02 Jan 2023 00:00:00 GMT",
  "digest": "SHA-256=RBNvo1WzZ4oRRq0W9+hknpT7T8If536DEMBg9hyq/4o=",
  "host": "remote.example.com",
  "signature": "keyId="https://myhost.example.com/example#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="neklhquVUYp5K2EN08PWvQS0AScZ2rGduHOno8hSG9k8xHW33Zm3r6UeKBJ3vs+RcSSnvatfcR/cnlfGBmDO0jbNCy71vKpzl0ExDKn5psHA8Tg1CmlH9wY4n00TnjYz1tTEFW9kBUp36yliYRlFDO+H2g25lCaGszD546yISO2rAZ/B3SF/pRWcNwuPZmJ+/As1VefUmTp94jTZ8fK5xjh0c4ulQTn3OQY1ppztaM029jqd9YlQR2rQMzgFc9+0VZPWarvLpzTh5qAwaNC18SRwhph5xXQy9tss4Zwpo39o+TcVUGgzMdj+1cZ3ogM6Tb1Unt7tcpkCtsDhww=="",
}
`;

exports[`signActivity 署名されたヘッダーを返す: default 1`] = `
{
  "date": "Sun, 01 Jan 2023 00:00:00 GMT",
  "digest": "SHA-256=RBNvo1WzZ4oRRq0W9+hknpT7T8If536DEMBg9hyq/4o=",
  "host": "remote.example.com",
  "signature": "keyId="https://myhost.example.com/example#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="CKZWtohqnHXjkIu3SEH20luhWcj6JdyBcPlPVoyeNJhRm9b7gJpIjUpCfCf7KsOR5r56tYHoMF2TG6ixJYOQ5NlCSRTBpOhmFfPL9ly2TS2O1wgm29qn6xFwqd4kTZn2NHxwR5BsMDjFfPD18sbxyhlkCSm8KbIExMJTIL+10FT8pQYgQXGZP6oGKQLQjgAp2SqvTEMrhUKZ5/2yFR5lDKp66RCarePhem78RDFePUh7QZSmlvK25G2IUlKs40KUmot4ZQD3yyEcteMXNJXXJ3PwfBTkqJshMbm/P9mYqOCfkUcmjEIT1qdP286qM/2Ewrac/6UCIQ5j0igGgg=="",
}
`;

exports[`signActivity 署名されたヘッダーを返す: path 1`] = `
{
  "date": "Sun, 01 Jan 2023 00:00:00 GMT",
  "digest": "SHA-256=RBNvo1WzZ4oRRq0W9+hknpT7T8If536DEMBg9hyq/4o=",
  "host": "remote.example.com",
  "signature": "keyId="https://myhost.example.com/example#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="cy1p2VbFR8S6RebKid90+gXdsb36LayujQI+cFIDg/blZm1eEmtPoub4d+PoyA18G1xLQ170G2txm0m1MKiM/+V3OxumwUWIpcF+eQP2vPZ9y3ZqFk12gjAmM8QEOwti72tuYT0Xtg19Ptag0N4iw62jp/3ntGSTsOvLbVkuFi53/CThiM/0XuF6Z5/C6Vthqd/5FS4TGLKNWb0KzDN8Rzrei0DbzCSYh2s5AUZhJq3Gx+JzwR1ppT1WIhqr3vSm8U6AC0DJJEQxYNwRYcWS9cPm3/oLz22tR5AMhvGV/ZRw+F+w54T1ciM4crav1xaCAP1zQ0+0Le5NB1lN0A=="",
}
`;

exports[`signActivity 署名されたヘッダーを返す: url 1`] = `
{
  "date": "Sun, 01 Jan 2023 00:00:00 GMT",
  "digest": "SHA-256=RBNvo1WzZ4oRRq0W9+hknpT7T8If536DEMBg9hyq/4o=",
  "host": "remote1.example.com",
  "signature": "keyId="https://myhost.example.com/example#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest",signature="Z7vtxdHR8DXUYEkvf1/Z1bxxzY7HAfiSihD22r1OQWCfKA+92aUmQX0rLySd4iI3kZKckhXk1DUz09poh5KEWLZl/x4MNjLKtjL7RjmuhodkNXamxz7HacWHIyAzl9M3fv+J2cwF1SKrey1EZndzYUn3wK4gDzIkyWCZZ9Q0zMkaLpqt5lFUnho4B1nVAUwgA1poneBoJZIKL+gkjHUr8Yrn/O6sk2g6VjhUqi5oMFDmI0/BVVwlSSPFt49QdAYevqRsZTUAkjObsH+rM4gPlhEduGCjteXdrwp+DIuVMvo3NE97hZEHmHkx4Op+78fM5w8eQj6loXCPANox9w=="",
}
`;
